src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/WeakEvent`1.cs

```diff
@@ -3,129 +3,63 @@
-                return weakHandlers
-                    .RemoveAll(WeakReferenceExtensions.IsNull)
-                    .Add(new WeakReference<EventHandler<TEventArgs>>(handler));
-            },
-            handler);
     }
 
     public void RemoveHandler(object target, EventHandler<TEventArgs> handler)
     {
-        lock (_keepAliveTable)
+        lock (_handlers.WriteLock)
         {
-            if (_keepAliveTable.TryGetValue(target, out var existingHandler))
+            if (_handlers.TryGetValue(target, out var existingHandler))
             {
                 var newHandler = existingHandler - handler;
-                if (newHandler is null)
+                if (newHandler != null)
                 {
-                    _keepAliveTable.Remove(target);
+                    _handlers.AddOrUpdate(target, newHandler);
                 }
                 else
                 {
-#if NET6_0_OR_GREATER
-                    _keepAliveTable.AddOrUpdate(target, newHandler);
-#else
-                    _keepAliveTable.Remove(target);
-                    _keepAliveTable.Add(target, newHandler);
-#endif
+                    _handlers.Remove(target);
                 }
             }
         }
-
-        ImmutableInterlocked.Update(
-            ref _weakHandlers,
-            static (weakHandlers, handler) =>
-            {
-                // Remove any references to the handler from the list of delegates to invoke for this event. During the
-                // mutation, also remove any handlers that were collected without being removed (and are now null).
-                var builder = weakHandlers.ToBuilder();
-                for (var i = weakHandlers.Count - 1; i >= 0; i--)
-                {
-                    var weakHandler = weakHandlers[i];
-                    if (!weakHandler.TryGetTarget(out var target))
-                    {
-                        // This handler was collected
-                        builder.RemoveAt(i);
-                        continue;
-                    }
-
-                    var updatedHandler = target - handler;
-                    if (updatedHandler is null)
-                    {
-                        builder.RemoveAt(i);
-                    }
-                    else if (updatedHandler != target)
-                    {
-                        builder[i].SetTarget(updatedHandler);
-                    }
-                }
-
-                return builder.ToImmutable();
-            },
-            handler);
     }
 
-    public void RaiseEvent(object? sender, TEventArgs e)
+    public void RaiseEvent(TEventArgs e)
     {
-        foreach (var weakHandler in _weakHandlers)
+        foreach (var (target, handler) in _handlers)
         {
-            if (!weakHandler.TryGetTarget(out var handler))
-                continue;
-
-            handler(sender, e);
+            handler(target, e);
```

Comment:
❗ `sender, e` was correct. `sender` and `target` are on opposite ends of this event, and for a weak event scenario are never the same instance.

Reply:
We want to send the `target` in this case though. Perhaps we could send both and not use EventHandler type.

Reply:
Added a new delegate type WeakEventHandler that takes both sender and target.